File: analyzer/src/io/github/colemakmods/keyboard/KeyboardAnalysis.java
package io.github.colemakmods.keyboard;

import io.github.colemakmods.chars.BigramFreq;
import io.github.colemakmods.chars.CharFreq;
import io.github.colemakmods.chars.FreqAnalysis;
import io.github.colemakmods.sa.SASettings;
import io.github.colemakmods.sa.SFBEffortModel;
import io.github.colemakmods.sa.SimpleEffortModel;
import io.github.colemakmods.sa.SimulatedAnnealing;

import java.io.*;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

/**
 * Perform an analysis on a keyboard layout using provided data on character and bigram frequencies,
 * penalties and finger configuration.

 * Created by steve on 18/10/14.
 */
public class KeyboardAnalysis {

    public static void main(String args[]) {
        if (args.length < 3) {
            exitHelp();
            return;
        }

        Options options = new Options();
        SASettings saSettings = null;
        List<CharFreq> charFreqs = null;
        List<BigramFreq> bigramFreqs = null;

        File keyboardFile = new File(args[args.length-1]);
        KeyboardLayout keyboardLayout = new KeyboardLayout(keyboardFile.getName());

        boolean valid = KeyboardMapping.parse(keyboardLayout, keyboardFile);
        if (!valid) return;
        keyboardLayout.dumpLayout(System.out);

        String alphabet = keyboardLayout.getAlphabet();

        for (int i=0; i<args.length-1; ++i) {
            if (args[i].equals("-c")) {
                String[] configFiles = args[++i].split(",");
                for (String configFile : configFiles) {
                    valid = KeyboardConfig.parse(keyboardLayout, new File(configFile));
                    if (!valid) return;
                }
                keyboardLayout.dumpConfig(System.out);

            } else if (args[i].equals("-f")) {
                String frequencyFile = args[++i];
                charFreqs = CharFreq.initialize(alphabet, new File(frequencyFile));
                bigramFreqs = BigramFreq.initialize(alphabet, new File(frequencyFile));
                if (charFreqs == null || bigramFreqs == null) {
                    return;
                }

            } else if (args[i].equals("-w")) {
                String wordFile = args[++i];

                FreqAnalysis fa = new FreqAnalysis(alphabet, -1);
                fa.analyze(new File(wordFile));

                charFreqs = fa.getCharFreqs();
                bigramFreqs = fa.getBigramFreqs();

            } else if (args[i].equals("-b")) {
                try {
                    String[] bigramsListSizes = args[++i].split(",");
                    if (bigramsListSizes.length >= 1) {
                        options.sfbListSize = Integer.parseInt(bigramsListSizes[0]);
                    }
                    if (bigramsListSizes.length >= 2) {
                        options.nfbListSize = Integer.parseInt(bigramsListSizes[1]);
                    }
                } catch (NumberFormatException ex) {
                    System.err.println("Invalid argument: bigram list size (-b)");
                    return;
                }

            } else if (args[i].startsWith("-o")) {
                options.outputFormat = args[++i];
            } else if (args[i].startsWith("--sa")) {
                String effortModelClass = args[++i];
                switch (effortModelClass) {
                    case "SimpleEffortModel":
                        saSettings = new SASettings(new SimpleEffortModel());
                        break;
                    case "SFBEffortModel":
                        saSettings = new SASettings(new SFBEffortModel());
                        break;
                    default:
                        System.err.println("Invalid argument: unknown --sa parameter");
                        return;
                }
            } else if (args[i].startsWith("-t")) {
                if (saSettings != null) {
                    saSettings.highTemp = Double.parseDouble(args[++i]);
                }
            } else if (args[i].startsWith("-a")) {
                if (saSettings != null) {
                    saSettings.alpha = Double.parseDouble(args[++i]);
                }
            }
        }

        List<String> errors = keyboardLayout.validate();
        if (!errors.isEmpty()) {
            for (String error : errors) {
                System.err.println("Error: " + error);
            }
            return;
        }

        if (charFreqs == null || bigramFreqs == null) {
            exitHelp();
            return;
        }

        if (saSettings != null) {
            keyboardLayout = SimulatedAnnealing.runSimulation(keyboardLayout, charFreqs, bigramFreqs, saSettings);
        }

        KeyboardAnalysis ka = new KeyboardAnalysis();
        LayoutResults layoutResults = ka.performAnalysis(keyboardLayout, charFreqs, bigramFreqs);

        try {
            // generate full text report
            if (options.outputFormat.indexOf('t') >= 0) {
                KeyboardAnalysisTextReport kr = new KeyboardAnalysisTextReport(options.sfbListSize, options.nfbListSize);
                String output = kr.generate(layoutResults);
                System.out.println(output);
            }
            // generate brief report for html table
            if (options.outputFormat.indexOf('b') >= 0) {
                KeyboardAnalysisBriefHTMLReport kr = new KeyboardAnalysisBriefHTMLReport();
                String output = kr.generate(layoutResults);
                System.out.println(output);
            }
        } catch (IOException ex) {
            ex.printStackTrace();
        }
    }

    private static void exitHelp() {
        System.out.println("KeyboardAnalysis  -c configFiles..  -f frequencyFile  keyboardFile");
        System.out.println("KeyboardAnalysis  -c configFiles..  -w wordFile  keyboardFile");
        System.out.println("KeyboardAnalysis  -c configFiles..  -w wordFile  --sa effortModel  keyboardStartingFile");
        System.exit(0);
    }

    public LayoutResults performAnalysis(KeyboardLayout keyboardLayout, List<CharFreq> charFreqs, List<BigramFreq> bigramFreqs) {
        List<String> messages = sanityCheck(keyboardLayout);

        HashMap<Key, Double> keyFreq = calculateKeyFrequency(keyboardLayout, charFreqs);
        double[] fingerFreq = calculateFingerFrequency(keyboardLayout, charFreqs);

        double handAlternation = countHandAlternation(keyboardLayout, bigramFreqs);

        //find difficult bigrams for the current layout
        List<FingerBigram> sameFingerBigrams = findSameFingerBigrams(keyboardLayout, bigramFreqs);
        List<FingerBigram> neighbourPenaltyFingerBigrams = findNeighbourFingerBigrams(keyboardLayout, bigramFreqs, false);
        List<FingerBigram> neighbourBenefitFingerBigrams = findNeighbourFingerBigrams(keyboardLayout, bigramFreqs, true);

        double[][] fingerEffort = new double[3][];
        fingerEffort[0] = calculateBaseEffort(keyboardLayout, charFreqs);
        fingerEffort[1] = calculateSameFingerBigramEffort(keyboardLayout, sameFingerBigrams);
        fingerEffort[2] = calculateNeighbourFingerBigramEffort(keyboardLayout,
                joinList(neighbourPenaltyFingerBigrams, neighbourBenefitFingerBigrams));

        return new LayoutResults(keyboardLayout, messages, keyFreq, fingerFreq, handAlternation,
            sameFingerBigrams, neighbourPenaltyFingerBigrams, neighbourBenefitFingerBigrams, fingerEffort);
    }

    private List<FingerBigram> joinList(List<FingerBigram> list1, List<FingerBigram> list2) {
        List<FingerBigram> list = new ArrayList<>();
        list.addAll(list1);
        list.addAll(list2);
        return list;
    }

    private List<String> sanityCheck(KeyboardLayout keyboardLayout) {
        List<String> messages = new ArrayList<>();

        String alphabet = keyboardLayout.getAlphabet();
        if (keyboardLayout.getRows() > 4) {
            messages.add("Warning: Too many rows");
        }

        for (char c = 'A'; c <= 'Z'; ++c) {
            if (alphabet.indexOf(c) < 0) {
                messages.add("Warning: letter " + c + " is missing from layout. ");
            }
        }
        for (int i=0; i<alphabet.length(); ++i) {
            char c = alphabet.charAt(i);
            if (alphabet.indexOf(c, i+1) >= 0) {
                messages.add("Warning: letter " + c + " appears more than once on layout. ");
            }
        }
        return messages;
    }

    private HashMap<Key, Double> calculateKeyFrequency(KeyboardLayout keyboardLayout, List<CharFreq> charFreqs) {
        HashMap<Key, Double> keyFrequency = new HashMap<>();
        for (int row = 0; row < keyboardLayout.getRows(); ++row) {
            for (int col = 0; col < keyboardLayout.getCols(); ++col) {
                Key key = keyboardLayout.lookupKey(row, col);
                if (key != null) {
                    Double freqTotal = 0.;
                    for (char ch : key.getChars().toCharArray()) {
                        CharFreq cf = CharFreq.findByChar(ch, charFreqs);
                        if (cf != null) freqTotal += cf.getFreq();
                    }
                    keyFrequency.put(key, freqTotal);
                }
            }
        }
        return keyFrequency;
    }

    private double[] calculateFingerFrequency(KeyboardLayout keyboardLayout, List<CharFreq> charFreqs) {
        double[] fingerFreq = new double[10];

        for (CharFreq charFreq : charFreqs) {
            char c = charFreq.getChar();
            Key key = keyboardLayout.lookupKey(c);
            if (key != null) {
                fingerFreq[key.getFinger()] += charFreq.getFreq();
            }
        }

        return fingerFreq;
    }

    private double countHandAlternation(KeyboardLayout keyboardLayout, List<BigramFreq> bigramFreqs) {
        double handAlternation = 0f;
        for (BigramFreq bigramFreq : bigramFreqs) {
            char[] chars = bigramFreq.getString().toCharArray();
            Key key1 = keyboardLayout.lookupKey(chars[0]);
            Key key2 = keyboardLayout.lookupKey(chars[1]);
            if (key1.getHand() != key2.getHand()) {
                handAlternation += bigramFreq.getFreq();
            }
        }
        return handAlternation;
    }

    /**
     * Return a list of all same-finger bigrams for this layout
     *
     * @param keyboardLayout the current keyboard layout
     * @param bigramFreqs the full set of bigram frequencies
     * @return the list of same-finger bigrams found
     */
    private List<FingerBigram> findSameFingerBigrams(KeyboardLayout keyboardLayout, List<BigramFreq> bigramFreqs) {
        List<FingerBigram> fingerBigrams = new ArrayList<FingerBigram>();

        for (BigramFreq bigramFreq : bigramFreqs) {
            char[] chars = bigramFreq.getString().toCharArray();
            Key key1 = keyboardLayout.lookupKey(chars[0]);
            Key key2 = keyboardLayout.lookupKey(chars[1]);
            if (key1 != null && key2 != null) {
                if (chars[0] != chars[1] && KeyboardConfig.isSameFinger(key1.getFinger(), key2.getFinger())) {
                    FingerBigram fingerBigram = new FingerBigram(key1, key2, bigramFreq);
                    fingerBigrams.add(fingerBigram);
                }

            }
        }

        return fingerBigrams;
    }

    /**
     * Return a list of neighbour-finger bigrams for this layout that have defined penalties
     *
     * @param keyboardLayout the current keyboard layout
     * @param bigramFreqs the full set of bigram frequencies
     * @param benefit whether to get beneficial or detrimental bigrams
     * @return the list of neighbour-finger bigrams found
     */
    private List<FingerBigram> findNeighbourFingerBigrams(KeyboardLayout keyboardLayout, List<BigramFreq> bigramFreqs, boolean benefit) {
        List<FingerBigram> fingerBigrams = new ArrayList<FingerBigram>();

        for (BigramFreq bigramFreq : bigramFreqs) {
            char[] chars = bigramFreq.getString().toCharArray();
            Key key1 = keyboardLayout.lookupKey(chars[0]);
            Key key2 = keyboardLayout.lookupKey(chars[1]);
            if (key1 != null && key2 != null) {
                if (KeyboardConfig.isNeighbourFinger(key1.getFinger(), key2.getFinger())) {
                    int outermostFinger = KeyboardConfig.getOutermostFinger(key1.getFinger(), key2.getFinger());
                    FingerBigram fingerBigram = new FingerBigram(key1, key2, bigramFreq);
                    if (keyboardLayout.hasPenaltyNeighbourFinger(outermostFinger) != benefit) {
                        fingerBigrams.add(fingerBigram);
                    }
                }

            }
        }

        return fingerBigrams;
    }

    /**
     * Calculate the base effort incurred due to key position
     */
    private double[] calculateBaseEffort(KeyboardLayout keyboardLayout, List<CharFreq> charFreqs) {
        double[] baseEffort = new double[10];
        for (char ch : keyboardLayout.getAlphabet().toCharArray()) {
            Key key = keyboardLayout.lookupKey(ch);
            CharFreq cf = CharFreq.findByChar(ch, charFreqs);
            if (cf != null) {
                baseEffort[key.getFinger()] += key.getEffort() * cf.getFreq();
            }
        }
        return baseEffort;
    }

    /**
     * Calculate the effort incurred due to same-finger bigram penalties
     */
    private double[] calculateSameFingerBigramEffort(KeyboardLayout keyboardLayout, List<FingerBigram> fingerBigrams) {
        double[] bigramEffort = new double[10];
        for (FingerBigram fingerBigram : fingerBigrams) {
            int finger1 = fingerBigram.getKey1().getFinger();
            int diffrows = Math.min(2, Math.abs(fingerBigram.getKey1().getRow() - fingerBigram.getKey2().getRow()));
            double penalty = keyboardLayout.getPenaltySameFinger(diffrows);
            bigramEffort[finger1] += fingerBigram.getBigramFreq().getFreq() * penalty;
        }
        return bigramEffort;
    }

    /**
     * Calculate the effort incurred due to neighbour-finger bigram penalties
     */
    private double[] calculateNeighbourFingerBigramEffort(KeyboardLayout keyboardLayout, List<FingerBigram> fingerBigrams) {
        double[] bigramEffort = new double[10];
        for (FingerBigram fingerBigram : fingerBigrams) {
            int finger1 = fingerBigram.getKey1().getFinger();
            int finger2 = fingerBigram.getKey2().getFinger();
            int outermostFinger = KeyboardConfig.getOutermostFinger(finger1, finger2);
            int diffrows = Math.min(2, Math.abs(fingerBigram.getKey1().getRow() - fingerBigram.getKey2().getRow()));
            double penalty = keyboardLayout.getPenaltyNeighbourFinger(outermostFinger, diffrows);
            bigramEffort[outermostFinger] += fingerBigram.getBigramFreq().getFreq() * penalty;
        }
        return bigramEffort;
    }

}

File: analyzer/src/io/github/colemakmods/keyboard/KeyboardAnalysis.kt
package io.github.colemakmods.keyboard

import io.github.colemakmods.chars.BigramDataLoader
import io.github.colemakmods.chars.BigramFreq
import io.github.colemakmods.chars.CharDataLoader
import io.github.colemakmods.chars.CharFreq
import io.github.colemakmods.chars.FreqAnalysis
import io.github.colemakmods.keyboard.report.KeyboardAnalysisBriefHTMLReport
import io.github.colemakmods.keyboard.report.KeyboardAnalysisTextReport
import io.github.colemakmods.sa.SASettings
import io.github.colemakmods.sa.SFBEffortModel
import io.github.colemakmods.sa.SimpleEffortModel
import io.github.colemakmods.sa.SimulatedAnnealing
import java.io.File
import java.io.IOException

/**
 * Perform an analysis on a keyboard layout using provided data on character and bigram frequencies,
 * penalties and finger configuration.
 *
 * Created by steve on 18/10/14.
 */
class KeyboardAnalysis {
    fun performAnalysis(
        keyboardLayout: KeyboardLayout,
        charFreqs: List<CharFreq>,
        bigramFreqs: List<BigramFreq>
    ): LayoutResults {
        val messages = sanityCheck(keyboardLayout)
        val keyFreq = calculateKeyFrequency(keyboardLayout, charFreqs)
        val fingerFreq = calculateFingerFrequency(keyboardLayout, charFreqs)
        val handAlternation = countHandAlternation(keyboardLayout, bigramFreqs)

        //find difficult bigrams for the current layout
        val sameFingerBigrams = findSameFingerBigrams(keyboardLayout, bigramFreqs)
        val neighbourPenaltyFingerBigrams =
            findNeighbourFingerBigrams(keyboardLayout, bigramFreqs, false)
        val neighbourBenefitFingerBigrams =
            findNeighbourFingerBigrams(keyboardLayout, bigramFreqs, true)
        val fingerEffort = arrayOf(
            calculateBaseEffort(keyboardLayout, charFreqs),
            calculateSameFingerBigramEffort(keyboardLayout, sameFingerBigrams),
            calculateNeighbourFingerBigramEffort(
                keyboardLayout,
                joinList(neighbourPenaltyFingerBigrams, neighbourBenefitFingerBigrams)
            )
        )
        return LayoutResults(
            keyboardLayout,
            messages,
            keyFreq,
            fingerFreq,
            handAlternation,
            sameFingerBigrams,
            neighbourPenaltyFingerBigrams,
            neighbourBenefitFingerBigrams,
            fingerEffort
        )
    }

    private fun joinList(list1: List<FingerBigram>, list2: List<FingerBigram>): List<FingerBigram> {
        val list: MutableList<FingerBigram> = ArrayList()
        list.addAll(list1)
        list.addAll(list2)
        return list
    }

    private fun sanityCheck(keyboardLayout: KeyboardLayout): List<String> {
        val messages: MutableList<String> = ArrayList()
        val alphabet = keyboardLayout.generateAlphabet()
        if (keyboardLayout.rows > 4) {
            messages.add("Warning: Too many rows")
        }
        messages.addAll(alphabet.findMissingLetters())
        messages.addAll(alphabet.findDuplicateLetters())
        return messages
    }

    private fun calculateKeyFrequency(
        keyboardLayout: KeyboardLayout,
        charFreqs: List<CharFreq>
    ): HashMap<Key, Double> {
        val keyFrequency = HashMap<Key, Double>()
        for (row in 0 until keyboardLayout.rows) {
            for (col in 0 until keyboardLayout.cols) {
                val key = keyboardLayout.lookupKey(row, col)
                if (key != null) {
                    var freqTotal = 0.0
                    for (ch in key.chars.toCharArray()) {
                        val cf = charFreqs.firstOrNull { ch == it.char }
                        if (cf != null) {
                            freqTotal += cf.freq.toDouble()
                        }
                    }
                    keyFrequency[key] = freqTotal
                }
            }
        }
        return keyFrequency
    }

    private fun calculateFingerFrequency(
        keyboardLayout: KeyboardLayout,
        charFreqs: List<CharFreq>
    ): DoubleArray {
        val fingerFreq = DoubleArray(10)
        for ((c, _, freq) in charFreqs) {
            val key = keyboardLayout.lookupKey(c)
            if (key != null) {
                fingerFreq[key.finger] += freq.toDouble()
            }
        }
        return fingerFreq
    }

    private fun countHandAlternation(
        keyboardLayout: KeyboardLayout,
        bigramFreqs: List<BigramFreq>
    ): Double {
        var handAlternation = 0.0
        for ((string, _, freq) in bigramFreqs) {
            val chars = string.toCharArray()
            val key1 = keyboardLayout.lookupKey(chars[0]) ?: continue
            val key2 = keyboardLayout.lookupKey(chars[1]) ?: continue
            if (key1.getHand() !== key2.getHand()) {
                handAlternation += freq.toDouble()
            }
        }
        return handAlternation
    }

    /**
     * Return a list of all same-finger bigrams for this layout
     *
     * @param keyboardLayout the current keyboard layout
     * @param bigramFreqs the full set of bigram frequencies
     * @return the list of same-finger bigrams found
     */
    private fun findSameFingerBigrams(
        keyboardLayout: KeyboardLayout,
        bigramFreqs: List<BigramFreq>
    ): List<FingerBigram> {
        val fingerBigrams = mutableListOf<FingerBigram>()
        for (bigramFreq in bigramFreqs) {
            val chars = bigramFreq.string.toCharArray()
            val key1 = keyboardLayout.lookupKey(chars[0])
            val key2 = keyboardLayout.lookupKey(chars[1])
            if (key1 != null && key2 != null) {
                if (chars[0] != chars[1] && KeyboardConfig.isSameFinger(key1.finger, key2.finger)) {
                    val fingerBigram = FingerBigram(key1, key2, bigramFreq)
                    fingerBigrams.add(fingerBigram)
                }
            }
        }
        return fingerBigrams
    }

    /**
     * Return a list of neighbour-finger bigrams for this layout that have defined penalties
     *
     * @param keyboardLayout the current keyboard layout
     * @param bigramFreqs the full set of bigram frequencies
     * @param benefit whether to get beneficial or detrimental bigrams
     * @return the list of neighbour-finger bigrams found
     */
    private fun findNeighbourFingerBigrams(
        keyboardLayout: KeyboardLayout,
        bigramFreqs: List<BigramFreq>,
        benefit: Boolean
    ): List<FingerBigram> {
        val fingerBigrams: MutableList<FingerBigram> = ArrayList()
        for (bigramFreq in bigramFreqs) {
            val chars = bigramFreq.string.toCharArray()
            val key1 = keyboardLayout.lookupKey(chars[0])
            val key2 = keyboardLayout.lookupKey(chars[1])
            if (key1 != null && key2 != null) {
                if (KeyboardConfig.isNeighbourFinger(key1.finger, key2.finger)) {
                    val outermostFinger =
                        KeyboardConfig.getOutermostFinger(key1.finger, key2.finger)
                    val fingerBigram = FingerBigram(key1, key2, bigramFreq)
                    if (keyboardLayout.hasPenaltyNeighbourFinger(outermostFinger) != benefit) {
                        fingerBigrams.add(fingerBigram)
                    }
                }
            }
        }
        return fingerBigrams
    }

    /**
     * Calculate the base effort incurred due to key position
     */
    private fun calculateBaseEffort(
        keyboardLayout: KeyboardLayout,
        charFreqs: List<CharFreq>
    ): DoubleArray {
        val baseEffort = DoubleArray(10)
        for (ch in keyboardLayout.generateAlphabet().asCharArray()) {
            val key = keyboardLayout.lookupKey(ch) ?: continue
            val cf = charFreqs.firstOrNull { ch == it.char }
            if (cf != null) {
                baseEffort[key.finger] += key.effort * cf.freq
            }
        }
        return baseEffort
    }

    /**
     * Calculate the effort incurred due to same-finger bigram penalties
     */
    private fun calculateSameFingerBigramEffort(
        keyboardLayout: KeyboardLayout,
        fingerBigrams: List<FingerBigram>
    ): DoubleArray {
        val bigramEffort = DoubleArray(10)
        for (fingerBigram in fingerBigrams) {
            val finger1 = fingerBigram.key1.finger
            val diffrows = Math.min(2, Math.abs(fingerBigram.key1.row - fingerBigram.key2.row))
            val penalty = keyboardLayout.getPenaltySameFinger(diffrows)
            bigramEffort[finger1] += fingerBigram.bigramFreq.freq * penalty
        }
        return bigramEffort
    }

    /**
     * Calculate the effort incurred due to neighbour-finger bigram penalties
     */
    private fun calculateNeighbourFingerBigramEffort(
        keyboardLayout: KeyboardLayout,
        fingerBigrams: List<FingerBigram>
    ): DoubleArray {
        val bigramEffort = DoubleArray(10)
        for (fingerBigram in fingerBigrams) {
            val finger1 = fingerBigram.key1.finger
            val finger2 = fingerBigram.key2.finger
            val outermostFinger = KeyboardConfig.getOutermostFinger(finger1, finger2)
            val diffrows = Math.min(2, Math.abs(fingerBigram.key1.row - fingerBigram.key2.row))
            val penalty = keyboardLayout.getPenaltyNeighbourFinger(outermostFinger, diffrows)
            bigramEffort[outermostFinger] += fingerBigram.bigramFreq.freq * penalty
        }
        return bigramEffort
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            if (args.size < 3) {
                exitHelp()
                return
            }
            val options = Options()
            var saSettings: SASettings? = null
            var charFreqs: List<CharFreq>? = null
            var bigramFreqs: List<BigramFreq>? = null
            val keyboardFile = File(args[args.size - 1])
            var keyboardLayout = KeyboardLayout(keyboardFile.name)
            var valid = KeyboardMapping.initialize(keyboardLayout, keyboardFile)
            if (!valid) return
            keyboardLayout.dumpLayout(System.out)
            val alphabet = keyboardLayout.generateAlphabet()
            val charDataLoader = CharDataLoader()
            val bigramDataLoader = BigramDataLoader()
            var i = 0
            while (i < args.size - 1) {
                if (args[i] == "-c") {
                    val configFiles =
                        args[++i].split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                    for (configFile in configFiles) {
                        valid = KeyboardConfig.initialize(keyboardLayout, File(configFile))
                        if (!valid) return
                    }
                    keyboardLayout.dumpConfig(System.out)
                } else if (args[i] == "-f") {
                    val frequencyFile = args[++i]
                    charFreqs = charDataLoader.initialize(alphabet, File(frequencyFile))
                    bigramFreqs = bigramDataLoader.initialize(alphabet, File(frequencyFile))
                    if (charFreqs == null || bigramFreqs == null) {
                        return
                    }
                } else if (args[i] == "-w") {
                    val wordFile = args[++i]
                    val fa = FreqAnalysis(alphabet, -1)
                    fa.analyze(File(wordFile))
                    charFreqs = fa.charFreqList
                    bigramFreqs = fa.bigramFreqList
                } else if (args[i] == "-b") {
                    try {
                        val bigramsListSizes =
                            args[++i].split(",".toRegex()).dropLastWhile { it.isEmpty() }
                                .toTypedArray()
                        if (bigramsListSizes.size >= 1) {
                            options.sfbListSize = bigramsListSizes[0].toInt()
                        }
                        if (bigramsListSizes.size >= 2) {
                            options.nfbListSize = bigramsListSizes[1].toInt()
                        }
                    } catch (ex: NumberFormatException) {
                        System.err.println("Invalid argument: bigram list size (-b)")
                        return
                    }
                } else if (args[i].startsWith("-o")) {
                    options.outputFormat = args[++i]
                } else if (args[i].startsWith("--sa")) {
                    val effortModelClass = args[++i]
                    saSettings = when (effortModelClass) {
                        "SimpleEffortModel" -> SASettings(SimpleEffortModel())
                        "SFBEffortModel" -> SASettings(SFBEffortModel())
                        else -> {
                            System.err.println("Invalid argument: unknown --sa parameter")
                            return
                        }
                    }
                } else if (args[i].startsWith("-t")) {
                    if (saSettings != null) {
                        saSettings.highTemp = args[++i].toDouble()
                    }
                } else if (args[i].startsWith("-a")) {
                    if (saSettings != null) {
                        saSettings.alpha = args[++i].toDouble()
                    }
                }
                ++i
            }
            val errors = keyboardLayout.validate()
            if (!errors.isEmpty()) {
                for (error in errors) {
                    System.err.println("Error: $error")
                }
                return
            }
            if (charFreqs == null || bigramFreqs == null) {
                exitHelp()
                return
            }
            if (saSettings != null) {
                keyboardLayout = SimulatedAnnealing.runSimulation(
                    keyboardLayout,
                    charFreqs,
                    bigramFreqs,
                    saSettings
                )
            }
            val ka = KeyboardAnalysis()
            val layoutResults = ka.performAnalysis(keyboardLayout, charFreqs, bigramFreqs)
            try {
                // generate full text report
                if (options.outputFormat.indexOf('t') >= 0) {
                    val kr = KeyboardAnalysisTextReport(options.sfbListSize, options.nfbListSize)
                    val output = kr.generate(layoutResults)
                    println(output)
                }
                // generate brief report for html table
                if (options.outputFormat.indexOf('b') >= 0) {
                    val kr = KeyboardAnalysisBriefHTMLReport()
                    val output = kr.generate(layoutResults)
                    println(output)
                }
            } catch (ex: IOException) {
                ex.printStackTrace()
            }
        }

        private fun exitHelp() {
            println("KeyboardAnalysis  -c configFiles..  -f frequencyFile  keyboardFile")
            println("KeyboardAnalysis  -c configFiles..  -w wordFile  keyboardFile")
            println("KeyboardAnalysis  -c configFiles..  -w wordFile  --sa effortModel  keyboardStartingFile")
            System.exit(0)
        }
    }
}
